File: legacy/pyqt-v2.3.0/core/average_service.py
from __future__ import annotations

from core.models import AverageCalculation


class AverageService:
    """Provides shared average-price calculations for the calculator and display modes."""

    def calculate(
        self,
        old_price: float,
        old_weight: float,
        new_price: float,
        new_weight: float,
    ) -> AverageCalculation:
        total_weight = old_weight + new_weight
        if total_weight <= 0:
            raise ValueError("total weight must be greater than zero")

        total_cost = (old_price * old_weight) + (new_price * new_weight)
        average_price = total_cost / total_weight
        return AverageCalculation(average_price=average_price, total_weight=total_weight)

File: src/pricing/average_service.cpp
#include "average_service.h"

#include <stdexcept>

namespace goldview {

AverageCalculation AverageService::calculate(
    double oldPrice,
    double oldWeight,
    double newPrice,
    double newWeight) const {
    const double totalWeight = oldWeight + newWeight;
    if (totalWeight <= 0.0) {
        throw std::invalid_argument("total weight must be greater than zero");
    }

    const double totalCost = (oldPrice * oldWeight) + (newPrice * newWeight);
    return AverageCalculation{totalCost / totalWeight, totalWeight};
}

std::wstring AverageService::formatHistoryRecord(
    double oldPrice,
    double oldWeight,
    double newPrice,
    double newWeight,
    const AverageCalculation& calculation) const {
    wchar_t buffer[160]{};
    swprintf_s(
        buffer,
        L"%.2f/%.2fg + %.2f/%.2fg -> %.2f",
        oldPrice,
        oldWeight,
        newPrice,
        newWeight,
        calculation.averagePrice);
    return buffer;
}

}  // namespace goldview
